server/src/main/java/org/elasticsearch/action/support/replication/TransportWriteAction.java

```diff
@@ -196,17 +208,28 @@ protected abstract void dispatchedShardOperationOnPrimary(
      */
     @Override
     protected void shardOperationOnReplica(ReplicaRequest request, IndexShard replica, ActionListener<ReplicaResult> listener) {
-        threadPool.executor(executorFunction.apply(replica)).execute(new ActionRunnable<>(listener) {
-            @Override
-            protected void doRun() {
-                dispatchedShardOperationOnReplica(request, replica, listener);
-            }
+        request.incRef();
+        final Releasable releasable = Releasables.releaseOnce(request::decRef);
+        boolean success = false;
+        try {
+            threadPool.executor(executorFunction.apply(replica)).execute(new ActionRunnable<>(
+                    ActionListener.runAfter(listener, releasable::close)) {
+                @Override
+                protected void doRun() {
+                    dispatchedShardOperationOnReplica(request, replica, listener);
+                }
 
-            @Override
-            public boolean isForceExecution() {
-                return true;
+                @Override
+                public boolean isForceExecution() {
+                    return true;
+                }
+            });
+            success = true;
+        } finally {
+            if (success == false) {
+                releasable.close();
```

Comment:
Can we also add `assert false` here since the runnable is forced to execute?

Reply:
++ added